twine/auth.py

```diff
@@ -0,0 +1,80 @@
+import warnings
+import getpass
+import functools
+
+import keyring
+import munch
+
+from . import utils
+from . import exceptions
+
+
+class Resolver:
+    def __init__(self, settings, username=None, password=None):
+        self.settings = settings
+        self.input = munch.Munch(username=username, password=password)
+
+    @property
+    @functools.lru_cache()
+    def username(self):
+        return utils.get_userpass_value(
+            self.input.username,
+            self.config,
+            key='username',
+            prompt_strategy=self.username_from_keyring_or_prompt,
+        )
+
+    @property
+    @functools.lru_cache()
+    def password(self):
+        return utils.get_userpass_value(
+            self.input.password,
+            self.config,
+            key='password',
+            prompt_strategy=self.password_from_keyring_or_prompt,
+        )
+
+    @property
+    def system(self):
+        return self.settings.system
+
+    @property
+    def config(self):
+        return self.settings.repository_config
+
+    def get_username_from_keyring(self):
+        try:
+            creds = keyring.get_credential(self.system, None)
+            if creds:
+                return creds.username
+        except AttributeError:
+            # To support keyring prior to 15.2
+            pass
+        except Exception as exc:
+            warnings.warn(str(exc))
+
+    def get_password_from_keyring(self):
+        try:
+            return keyring.get_password(self.system, self.username)
+        except Exception as exc:
+            warnings.warn(str(exc))
+
+    def username_from_keyring_or_prompt(self):
+        return (
+            self.get_username_from_keyring()
+            or self.prompt('username', input)
+        )
+
+    def password_from_keyring_or_prompt(self):
+        return (
+            self.get_password_from_keyring()
+            or self.prompt('password', getpass.getpass)
+        )
+
+    def prompt(self, what, how=None):
+        return how(f"Enter your {what}: ")
+
+
+class Private(Resolver):
+    def prompt(self, what, how=None):
+        raise exceptions.NonInteractive(f"Credential not found for {what}.")
```

Comment:
Instead of this subclass, what do you think about passing `non_interactive` as an arg to `Resolver`, then using `self.non_interactive` to control the prompt?

Reply:
I'd begun with just that approach, but then found this approach superior, especially as the goal of this PR is to reduce the number of places that a `non_interactive` parameter is passed/used. Passing it in would violate the goal:

> the 'non_interactive' boolean parameter now is passed into one function and causes its effect there

I particularly avoid boolean values, especially ones that switch behavior for an entire function, because they entangle the behavior with the decision logic. This proposed approach helps keep the decision separate from the behavior. It allows one to reason about a `Resolver` or `Private (resolver)` with less state. It makes it obvious that if you have a `Private` instance, you'll never be prompted, but if you have a `Resolver`, you'll always be prompted... and otherwise, it's guaranteed there's no other interaction with the boolean attribute.

The downside to the proposed approach is that subclass hierarchies don't scale well if you have other dimensions on which you wish to specialize, but my instinct is that such a change is unlikely and if it does arise, this code can once again be refactored to manage the switch in the object.

Reply:
In 098f106, I've added a helper classmethod that makes the code in settings cleaner.